store/src/java/com/zimbra/cs/service/mail/ModifyCalendarItem.java

```diff
@@ -108,49 +109,49 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
-            int modSeq = (int) request.getAttributeLong(MailConstants.A_MODIFIED_SEQUENCE, 0);
-            int revision = (int) request.getAttributeLong(MailConstants.A_REVISION, 0);
-            if (modSeq != 0 && revision != 0 &&
-                    (modSeq < calItem.getModifiedSequence() || revision < calItem.getSavedSequence())) {
-                throw MailServiceException.INVITE_OUT_OF_DATE(iid.toString());
-            }
-
-            Invite inv = calItem.getInvite(iid.getSubpartId(), compNum);
-            if (inv == null) {
-                throw MailServiceException.INVITE_OUT_OF_DATE(iid.toString());
-            }
-            Invite seriesInv = calItem.getDefaultInviteOrNull();
-            int folderId = calItem.getFolderId();
-            if (!isInterMboxMove && iidFolder != null) {
-                folderId = iidFolder.getId();
-            }
-            modifyCalendarItem(zsc, octxt, request, acct, mbox, folderId, calItem, inv, seriesInv,
-                               response, isInterMboxMove, sendQueue);
-        } finally {
-            mbox.lock.release();
-            sendQueue.send();
-        }
+		try (final MailboxLock l = mbox.lockFactory.writeLock()) {
+			l.lock();
+			CalendarItem calItem = mbox.getCalendarItemById(octxt, iid.getId());
+			if (calItem == null) {
+				throw MailServiceException.NO_SUCH_CALITEM(iid.toString(), "Could not find calendar item");
+			}
+
+			// Reject the request if calendar item is under trash or is being moved to
+			// trash.
+			if (calItem.inTrash()) {
+				throw ServiceException.INVALID_REQUEST("cannot modify a calendar item under trash", null);
+			}
+			if (!isInterMboxMove && iidFolder != null) {
+				if (iidFolder.getId() != calItem.getFolderId()) {
+					Folder destFolder = mbox.getFolderById(octxt, iidFolder.getId());
+					if (destFolder.inTrash()) {
+						throw ServiceException.INVALID_REQUEST("cannot combine with a move to trash", null);
+					}
+				}
+			}
+
+			// Conflict detection. Do it only if requested by client. (for backward compat)
+			int modSeq = (int) request.getAttributeLong(MailConstants.A_MODIFIED_SEQUENCE, 0);
+			int revision = (int) request.getAttributeLong(MailConstants.A_REVISION, 0);
+			if (modSeq != 0 && revision != 0
+					&& (modSeq < calItem.getModifiedSequence() || revision < calItem.getSavedSequence())) {
+				throw MailServiceException.INVITE_OUT_OF_DATE(iid.toString());
+			}
+
+			Invite inv = calItem.getInvite(iid.getSubpartId(), compNum);
+			if (inv == null) {
+				throw MailServiceException.INVITE_OUT_OF_DATE(iid.toString());
+			}
+			Invite seriesInv = calItem.getDefaultInviteOrNull();
+			int folderId = calItem.getFolderId();
+			if (!isInterMboxMove && iidFolder != null) {
+				folderId = iidFolder.getId();
+			}
+			modifyCalendarItem(zsc, octxt, request, acct, mbox, folderId, calItem, inv, seriesInv, response,
+					isInterMboxMove, sendQueue);
+		} finally {
+			sendQueue.send();
```

Comment:
It's holding the write lock while calling send() now.

Reply:
mvoed to another branch and recreated your comment there